File: internal/sorting/sorting.go
package sorting

func BubbleSort(numbers []int) {

	for i := range numbers {
		for j := 0; j < len(numbers)-i-1; j += 1 {
			if numbers[j] > numbers[j+1] {
				swap(j, j+1, numbers)
			}
		}
	}

}

func swap(idx1, idx2 int, numbers []int) {
	temp := numbers[idx1]

	numbers[idx1] = numbers[idx2]
	numbers[idx2] = temp
}

File: javascript/sorting/sorting.js


function bubble_sort(numbers) {
    for (let i=0; i < numbers.length; i++) {
        for (let j=0; j < numbers.length - i; j++) {
            if (numbers[j+1] < numbers[j]) {
                swap(j, j+1, numbers)
            }
        }
    }

}

function qs(arr, lo, hi) {
    if (lo >= hi) return

    const pivotIdx = partition(arr, lo, hi)

    qs(arr, lo, pivotIdx - 1)
    qs(arr, pivotIdx + 1, hi)
}

function partition(arr, lo, hi) {
    const pivot = arr[hi]
    let idx = lo - 1

    for (let i = lo; i < hi; i++) {
        const n = arr[i]
        if (n <= pivot) {
            idx++
            swap(i, idx, arr)
        }
    }

    idx++
    arr[hi] = arr[idx]
    arr[idx] = pivot

    return idx
}

function quick_sort(arr) {
    qs(arr, 0, arr.length - 1)
}

function swap(idx1, idx2, arr) {
    const temp = arr[idx1]
    arr[idx1] = arr[idx2]
    arr[idx2] = temp
}

module.exports = {
    bubble_sort,
    quick_sort
}


